src/main/java/com/gtnewhorizon/gtnhlib/capability/ICapabilityProvider.java

```diff
@@ -0,0 +1,51 @@
+package com.gtnewhorizon.gtnhlib.capability;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import net.minecraftforge.common.util.ForgeDirection;
+
+// spotless:off
+/**
+ * Classes implementing this interface indicates that it is able to offer certain {@link Capability}.
+ * Implementing this interface allows other systems to query and retrieve capabilities without relying on direct
+ * interface implementations. This can be useful for extending or overriding behavior at runtime.
+ * Typically implemented by TileEntities, Items or Entities.
+ * <p>
+ * Example usage:
+ * <pre>{@code
+ *     public class MyTileEntity implements ICapabilityProvider {
+ *         private MyCapabilityImplementation myCapability = new MyCapabilityImplementation();
+ *
+ *         @Override
+ *         public <T> T getCapability(@Nonnull Capability<T> capability, @Nullable ForgeDirection side) {
+ *             if (capability == MyCapability.CAPABILITY) {
+ *                 return MyCapability.CAPABILITY.cast(myCapability);
+ *             }
+ *             return null;
+ *         }
+ *     }
+ * }</pre>
+ */
+// spotless:on
+public interface ICapabilityProvider {
```

Comment:
This is worded pretty generically but as written this only makes sense for tile entities, which have sides. I would like to use capabilities for Covers and probably other things which might not have sides at a later time. Should they all be burdened with the concept of side? Should this be `ISidedCapabilityProvider` while `ICapabilityProvider` wouldn't have sides?

Perhaps not but I'd like to hear to what extent capabilities for non-tile-entities are planned to be supported.

In fact, could we wait to merge this until this includes an example of a capability for something  other than a tile entity?

Reply:
`side` can be `UNKNOWN` if 1. Queried capability doesn't have concept of sidedness 2. Queried capability is about TileEntity, but sidedness is not the interest. It's technically possible to separate into the two, but having two methods for providing capability sounds confusing to me.
Actually I tried to implement generics "context" rather than just `ForgeDirection`, just like seen in NeoForge. However, I found it difficult to implement cleanly due to the difference of how capability is registered. Moreover, I skimmed through popular repositories and found only `Void` and `Direction` being used, meaning generic context is useless, at least at this time.